### PR TITLE
Handle partial postcodes in admin user search

### PR DESCRIPTION
**Pull request by axlewin**
If a valid first part of a postcode (e.g. CB3) is provided, treat the centroid of this region as the postcode location.

### CODE REVIEW

File: src/main/java/uk/ac/cam/cl/dtg/util/locations/PostCodeIOLocationResolver.java [hunk 1]
```diff
@@ -185,40 +186,69 @@ private List<PostCode> carryOutExternalPostCodeServiceRequest(final List<String>
...
+        for (String postCode : unknownPostCodes) {
+            if (postCode.length() > 4) {
+                completePostCodes.add(postCode);
+            }
+            else {
+                outCodes.add(postCode);
+            }
+        }
+
+        if (completePostCodes.size() > POSTCODEIO_MAX_REQUESTS) {
```
Comment (jacbn) [inline src/main/java/uk/ac/cam/cl/dtg/util/locations/PostCodeIOLocationResolver.java:202]: We should certainly be checking the number of outCodes against this too.
```suggestion
        if (completePostCodes.size() + outCodes.size() > POSTCODEIO_MAX_REQUESTS) {
```

**Review by jacbn — COMMENTED** [review]
Looks good. I am not a fan of sending many individual requests but there really is no bulk endpoint equivalent so 🤷. We do need to fix the `MAX_REQUESTS` check though.

File: src/main/java/uk/ac/cam/cl/dtg/util/locations/PostCodeIOLocationResolver.java (hunk 1 above)
Comment (axlewin) [reply src/main/java/uk/ac/cam/cl/dtg/util/locations/PostCodeIOLocationResolver.java:202]: Given that `POSTCODEIO_MAX_REQUESTS` is based on the max. number of postcodes that can be batch-requested from the `/postcodes` endpoint in one request, comparing the combined number of postcodes and outcodes to this limit feels fairly arbitrary - but there should definitely be a limit, and this works as well as any.

Comment (jacbn) [reply src/main/java/uk/ac/cam/cl/dtg/util/locations/PostCodeIOLocationResolver.java]: Ah, okay. From a user perspective this keeps things consistent with how they were before which I prefer, though admittedly I struggle to see anyone using 100 different postcodes in one request anyway.

**Review by jacbn — APPROVED** [review]

**Pull request merged**